_blog/2022-12-23-procedural-macros.mdx

```diff
@@ -0,0 +1,417 @@
+
+One way to catch property errors ahead of time is if we had used an extension to JS like TypeScript, which can type the parameters of the function and check the arguments of callers.
+## Writing a Rust procedural macro
+If you are not in a Rust project/workspace already you can create one with the `cargo new *name*` command. Before we start generating code we should declare a trait as a target for our macros output.
+### The BinarySerializable trait
+In our current create, we need to create a `trait` that will be describe the requirements for serializing and deserializing objects. It will be called `BinarySerializable`. Serialization will require adding data to a buffer. Deserialization will require pulling from an iterator (which iterates over bytes of a serialized buffer) and producing a term of `Self`. *We will assume the format is well formed and panic at runtime rather than proper handling with `Result` when deserializing.*
+```rust
+pub trait BinarySerializable {
+    fn serialize(self, buf: &mut Vec<u8>);
+
+    fn deserialize<I: Iterator<Item = u8>>(iter: &mut I) -> Self;
+}
+```
+
+We can implement the `BinarySerializable` trait for the primitives that will be in our structures:
+```rust
+impl BinarySerializable for bool {
+    fn serialize(self, buf: &mut Vec<u8>) {
+        buf.push(self as u8)
+    }
+
+    fn deserialize<I: Iterator<Item = u8>>(iter: &mut I) -> Self {
+        iter.next().unwrap() == (true as u8)
+    }
+}
+
+impl BinarySerializable for u64 {
+    fn serialize(self, buf: &mut Vec<u8>) {
+        buf.extend_from_slice(&self.to_le_bytes());
+    }
+
+    fn deserialize<I: Iterator<Item = u8>>(iter: &mut I) -> Self {
+        let mut buf = [0; u64::BITS as usize / 8];
+        buf.fill_with(|| iter.next().unwrap());
+        u64::from_le_bytes(buf)
+    }
+}
+
+impl BinarySerializable for String {
+    fn serialize(self, buf: &mut Vec<u8>) {
+        (self.len() as u64).serialize(buf);
+        buf.extend_from_slice(self.as_bytes());
+    }
+
+    fn deserialize<I: Iterator<Item = u8>>(iter: &mut I) -> Self {
+        let length = u64::deserialize(iter) as usize;
+        String::from_utf8(iter.take(length).collect()).unwrap()
+    }
+}
+```
+
+> Again this is how to article write not low level byte manipulation. But I know which language I prefer for doing this logic 😄
+### The problem
+Now we have this, we want to add this to a definition for objects
+```rust
+struct Country {
+	name: String,
+	population: u64,
+	above_equator: bool
+}
+```
+
+We could implement `BinarySerializable` for `Country` writing manually in the following way:
+```rust
+impl BinarySerializable for Country {

```

Comment:
I think we should state it more explicitly that the macro will generate this code block

Reply:
Good idea 👍 "The derive attribute is transformed into the impl block"